dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala

```diff
@@ -544,23 +548,62 @@ class DistriOptimizer[T: ClassTag] (
     models = DistriOptimizer.initThreadModels(
       model, dataset, criterion, state, nodeNumber, coresPerNode, checkSingleton, parameters)
 
-
-    DistriOptimizer.optimize(
-      dataset,
-      coresPerNode,
-      state,
-      endWhen,
-      metrics,
-      models,
-      optimMethod,
-      parameters,
-      validationTrigger,
-      validationDataSet,
-      validationMethods,
-      checkpointTrigger,
-      checkpointPath,
-      isOverWrite
-    )
+    var retryNum = 0
+    val maxRetry = System.getProperty("bigdl.failure.retryTimes", "5").toInt
+    var lastFailureTimestamp = System.nanoTime()
+    while (retryNum < maxRetry) {
+      try {
+        DistriOptimizer.optimize(
+          dataset,
+          coresPerNode,
+          state,
+          endWhen,
+          metrics,
+          models,
+          optimMethod,
+          parameters,
+          validationTrigger,
+          validationDataSet,
+          validationMethods,
+          checkpointTrigger,
+          checkpointPath,
+          isOverWrite
+        )
+      } catch {
+        case t: Throwable =>
+          DistriOptimizer.logger.error("Error: " + ExceptionUtils.getStackTrace(t))
+          if (checkpointPath.isDefined) {
+            if (System.nanoTime() - lastFailureTimestamp < maxRetry * 2 * 60 * 1e9) {
```

Comment:
It's used to make sure in a long time training, each exception is able to rerun maxRetry times.  Will add the explaination later in the comments.